### PR TITLE
[TRTLLM-15788][feat] Add dflash 2 support

### PR DESCRIPTION
**Pull request by mikeiovine**
## Description

Support DFlash 2 (https://huggingface.co/incoai/Qwen3.8-27B-DFlash2). DFlash 2 has a new drafting architecture which should improve acceptance rates.

See this blog for more details: https://inco.ai/blog/dflash2/

## Test Coverage

New unit tests. 

Will add more end to end tests when we get the drafter into the shared model scratch space.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Added DFlash 2 support for candidate-path scoring, checkpoint-based configuration, vocabulary validation, tensor-parallel logits, and downstream sampling.
- Added Hub model checkpoint resolution while preserving local model behavior.
- Added DFlash 2 documentation and clarified related implementation comments.
- Review focus: validate tensor-parallel logits handling, checkpoint compatibility, error paths, performance, and API consistency.
- No configuration-file or test-list changes were identified.

## QA Engineer Review

- Added DFlash 2 semantic tests for candidate selection, logits handling, configuration validation, masks, checkpoint contracts, fallback behavior, block decoding, data types, malformed weights, vocabulary mapping, and drafter compatibility.
- Updated the DFlash mask-wrapper test helper to support DFlash 2 configurations.
- Test-list coverage was not provided for the changed test code.
- Verdict: needs follow-up.

### CODE REVIEW

**Comment by mikeiovine** [comment]
Adding full pre-merge approved tag as only failure is infra (failed to reserve node)

**Review by arysef — APPROVED** [review]

File: tensorrt_llm/llmapi/llm_utils.py [hunk 1]
```diff
@@ -419,8 +419,19 @@ def __call__(self) -> Tuple[Path, Union[Path, None]]:
                 self.llm_args.speculative_config.speculative_model is not None):
             spec_model_obj = _ModelWrapper(
                 self.llm_args.speculative_config.speculative_model)
+            was_hub_model = spec_model_obj.is_hub_model
             spec_model_dir = self._download_hf_model_if_needed(spec_model_obj)
             self.llm_args.speculative_config.speculative_model = spec_model_dir
+            # Some speculative configs read defaults out of the draft
+            # checkpoint's config.json (e.g. DFlash's target_layer_ids). A hub
+            # repo id had no readable config.json at validation time; a local
+            # path was already resolved back then.
```
Comment (mikeiovine) [inline tensorrt_llm/llmapi/llm_utils.py:428]: Note for reviewers; this handles a specific case where the drafter isn't already on disk. Honestly, the right fix is probably to get rid of the `config.json` reads in `DecodingConfig` initialization. Leaving in for now to fix this issue; can fix properly in a follow up after this lands.

File: tensorrt_llm/_torch/models/modeling_dflash.py [hunk 2]
```diff
@@ -677,16 +1068,29 @@ def _get_attention_mask_args(self, layer_idx):
...
         if sliding_window is None:
             # Legacy drafters without an explicit window preserve their prior
             # non-windowed behavior.
             return causal, (-1, -1)
         # FlashAttention's bounds are inclusive: W tokens are current + W-1 left.
-        return causal, (sliding_window - 1, 0)
+        # A non-causal sliding layer windows both sides, matching the HF flash
+        # path these drafters are trained under (see dspark_layer_window_size).
+        if not causal:
+            return False, (sliding_window - 1, sliding_window - 1)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_dflash.py:1105]: The symmetric right window returned here is only honored by the VANILLA backend — the TRTLLM path passes just `window_left = window_size[0]` to `batch_decode_with_kv_cache`/`batch_context_with_kv_cache` (modeling_dflash.py:1428), silently dropping `window_size[1]`. Within a block, keys sit at most `block_size - 1` positions to a query's right, so this only diverges when `sliding_window < block_size` — impossible for the released drafter (window 2048), but `block_size = max_draft_len + 1` is user-settable, and the failure mode is silent accuracy loss with backend-dependent results. Suggest raising (or at least warning) when a non-causal windowed layer meets `sliding_window < block_size` on the TRTLLM backend.

File: tensorrt_llm/_torch/speculative/dflash.py [hunk 3]
```diff
@@ -1140,6 +1153,95 @@ def _refine_block_logits(
...
+        position's top-k candidates, scored under the predecessor the walk
+        settled on. Downstream sampling is unchanged and stays lossless: each
+        row is the distribution its draft token was drawn from.
+
+        The walk always chains through the greedy successor, so under sampling
+        a draft token can differ from the predecessor the next position was
+        scored under. That costs acceptance rate, not correctness — same
+        trade-off as ``_apply_dspark_markov_bias``.
+        """
+        if self._d2t is not None:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/speculative/dflash.py]: This d2t incompatibility only surfaces at the first generation step, after the model has loaded and started serving. Both `_d2t` and `has_candidate_selector` are known at worker setup / weight-load time — checking there would fail fast at init instead of mid-request. (Same applies to the unsupported-shard-width `NotImplementedError` below, though that one depends on the runtime logits shape and is harder to hoist.)

File: tensorrt_llm/_torch/speculative/dflash.py [hunk 4]
```diff
@@ -1140,6 +1153,95 @@ def _refine_block_logits(
...
+                f"logits of width {shard}."
+            )
+
+        from ..distributed.ops import allgather
+
+        local_values, local_ids = torch.topk(gen_logits, top_k, dim=-1)
+        local_ids = local_ids + mapping.tp_rank * shard
+        # Interleaved (id, value) pairs, matching _get_local_max_and_combined.
+        combined = torch.stack([local_ids.float(), local_values.float()], dim=-1).flatten(-2)
+        gathered = allgather(combined, mapping, dim=-1)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/speculative/dflash.py:1334]: The TP shard-reduction path is only exercised against a mocked `allgather` (the unit test replays the reduction in-process), so the real-collective properties it depends on — rank-order concatenation along `dim=-1` and the interleaved-pair parity surviving the concat (each rank contributes an even `2 * top_k` chunk, which holds) — are unverified on hardware. A small 2-GPU test alongside the planned e2e runs would close this; until then TP>1 DFlash 2 should be treated as untested.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

Two things the description doesn't mention that reach beyond DFlash 2:

- `[llm_utils.py:429](https://github.com/NVIDIA/TensorRT-LLM/pull/18155/files#diff-5a12dda2ba2e2e400735911315f798c3e1bbdf9bdbe518d038d0d1d8f26980f2R429)` — re-resolving `resolve_from_checkpoint` after a hub download fixes HF-repo-id draft checkpoints for **plain DFlash too** (previously `target_layer_ids`/`mask_token_id` were never filled for hub models). Worth a line in the description since it changes behavior for existing users.
- `[llm_args.py:3057](https://github.com/NVIDIA/TensorRT-LLM/pull/18155/files#diff-bf140a489175e4c6e5931cd19cc57e06c1634483517f06c4e6fda96d5d26b056R3057)` — the new `block_size` vs `max_draft_len` warning also fires for DFlash 1 checkpoints on every validation pass. Fine, but it's a user-visible behavior change outside the "DFlash 2 support" framing.

The core math (grouped conv, edge scoring, lattice walk) is verified against nested-loop oracles including the request-boundary and anchor-slot edge cases, plus a bf16-negative-control block-decode parity test — that's the right level of coverage for a new drafting algorithm. Remaining gaps are the real multi-rank collective (the TP test mocks `allgather`) and end-to-end acceptance-rate runs, which the description already flags as pending. Suggest tracking both explicitly so they don't get lost once the drafter lands in the model scratch space.

**Review by lori-ren — APPROVED** [review]

**Review by Wanli-Jiang — APPROVED** [review]

File: tensorrt_llm/_torch/speculative/dflash.py (hunk 4 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/speculative/dflash.py:1334]: I'll add these tests in a follow up. We need to add the dflash 2 drafter to the llm models repo anyways; was planning on adding more tests once that's done.

File: tensorrt_llm/_torch/speculative/dflash.py (hunk 3 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/speculative/dflash.py]: This one can probably just be removed entirely, to be honest.

File: tensorrt_llm/_torch/models/modeling_dflash.py (hunk 2 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/models/modeling_dflash.py:1105]: Going to add this validation with the next PR/extra tests to avoid another CI run
